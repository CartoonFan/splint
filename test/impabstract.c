typedef int mint;
typedef /*@concrete@*/ int cint;

int f (mint m, cint c)
{
  if (c > 3)
    {
      return c;
    }

  if (m < 2)
    {
      return m;
    }

  return 3;
}
