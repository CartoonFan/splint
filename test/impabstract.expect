
Finished LCLint checking --- no code errors found

impabstract.c: (in function f)
impabstract.c:11:7: Left operand of < is abstract type (mint): m < 2
  An abstraction barrier is broken. If necessary, use /*@access <type>@*/ to
  allow access to an abstract type. (-abstract will suppress message)
impabstract.c:13:14: Return value type mint does not match declared type int: m
  Underlying types match, but mint is an abstract type that is not accessible
  here.

Finished LCLint checking --- 2 code errors found, as expected

impabstract.c:2:28: Mutable abstract type cint declared without pointer
                       indirection: int (violates assignment semantics)
  LCL semantics requires that a mutable type exhibits sharing semantics. In
  order for objects to be shared a indirection is necessary in the
  representation. A mutable type may be represented by a pointer or an abstract
  mutable type. Handles into static data are fine, too, but will generate this
  error message unless it is suppressed. (-mutrep will suppress message)
impabstract.c:2:28: Datatype cint inconsistently declared as concrete type
  A function, variable or constant is redefined with a different type.
  (-incondefs will suppress message)
   impabstract.lcl:2:1: Specification of cint
impabstract.c: (in function f)
impabstract.c:6:7: Left operand of > is abstract type (cint): c > 3
  An abstraction barrier is broken. If necessary, use /*@access <type>@*/ to
  allow access to an abstract type. (-abstract will suppress message)
impabstract.c:8:14: Return value type cint does not match declared type int: c
  Underlying types match, but cint is an abstract type that is not accessible
  here.

Finished LCLint checking --- 4 code errors found, as expected
