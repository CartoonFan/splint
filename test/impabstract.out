
Cannot find standard library: ansi.lcd
     Check LARCH_PATH environment variable.

Finished LCLint checking --- no code errors found

Cannot find standard library: ansi.lcd
     Check LARCH_PATH environment variable.
impabstract.c: (in function f)
impabstract.c:11:7: Left operand of < is abstract type (mint): m < 2
  An abstraction barrier is broken. If necessary, use /*@access <type>@*/ to
  allow access to an abstract type. (-abstract will suppress message)
impabstract.c:13:14: Return value type mint does not match declared type int: m
  Underlying types match, but mint is an abstract type that is not accessible
  here.

Finished LCLint checking --- 2 code errors found, as expected

Cannot find standard library: ansi.lcd
     Check LARCH_PATH environment variable.
Continuing without LCL init file: lclinit.lci
Continuing without LSL init file: lslinit.lsi
Unable to find CTrait.syms or CTrait.lsl.  Check LARCH_PATH environment
variable.
